include/RAJA/pattern/detail/reduce.hpp

```diff
@@ -0,0 +1,397 @@
+};
+
+#ifdef RAJA_ENABLE_TARGET_OPENMP
+#pragma omp end declare target
+#endif
+
+namespace detail
+{
+
+template <typename T, bool doing_min = true>
+class ValueLoc
+{
+public:
+  T val = doing_min ? operators::limits<T>::max() : operators::limits<T>::min();
+  Index_type loc = -1;
+
+  constexpr ValueLoc() = default;
+  constexpr ValueLoc(ValueLoc const &) = default;
+  ValueLoc &operator=(ValueLoc const &) = default;
+  RAJA_HOST_DEVICE constexpr ValueLoc(T const &val) : val{val}, loc{-1} {}
+  RAJA_HOST_DEVICE constexpr ValueLoc(T const &val, Index_type const &loc)
+      : val{val}, loc{loc}
+  {
+  }
+
+  RAJA_HOST_DEVICE operator T() const { return val; }
+  RAJA_HOST_DEVICE Index_type getLoc() { return loc; }
+  RAJA_HOST_DEVICE bool operator<(ValueLoc const &rhs) const
+  {
+    return val < rhs.val;
+  }
+  RAJA_HOST_DEVICE bool operator>(ValueLoc const &rhs) const
+  {
+    return val > rhs.val;
+  }
+};
+
+}  // end detail
+
+}  // end reduce
+
+namespace operators
+{
+template <typename T, bool B>
+struct limits<::RAJA::reduce::detail::ValueLoc<T, B>> : limits<T> {
+};
+}
+
+namespace reduce
+{
+
+namespace detail
+{
+
+template <typename T,
+          template <typename> class Reduce_,
+          template <typename, typename> class Combiner_>
+class BaseReduce
+{
+  using Reduce = Reduce_<T>;
+  // NOTE: the _t here is to appease MSVC
+  using Combiner_t = Combiner_<T, Reduce>;
+  Combiner_t mutable c;
+
+public:
+  using value_type = T;
+  using reduce_type = Reduce;
+
+  //! prohibit compiler-generated default ctor
+  BaseReduce() = delete;
+
+  //! prohibit compiler-generated copy assignment
+  BaseReduce &operator=(const BaseReduce &) = delete;
+
+  //! compiler-generated copy constructor
+  BaseReduce(const BaseReduce &) = default;
+
+  //! compiler-generated move constructor
+  BaseReduce(BaseReduce &&) = default;
+
+  //! compiler-generated move assignment
+  BaseReduce &operator=(BaseReduce &&) = default;
+
+  constexpr BaseReduce(T init_val, T identity_ = Reduce::identity())
+      : c{init_val, identity_}
+  {
+  }
+
+  void combine(T const &other) const { c.combine(other); }
+
+  T &local() const { return c.local(); }
+
+  //! Get the calculated reduced value
+  operator T() const { return c.get(); }
+
+  //! Get the calculated reduced value
+  T get() const { return c.get(); }
+};
+
+template <typename T, typename Reduce, typename Derived>
+class BaseCombinable
+{
+protected:
+  BaseCombinable const *parent = nullptr;
+  T identity;
+  T mutable my_data;
+
+public:
+  //! prohibit compiler-generated default ctor
+  BaseCombinable() = delete;
+
+  constexpr BaseCombinable(T init_val, T identity_ = T())
+      : identity{identity_}, my_data{init_val}
+  {
+  }
+
+  constexpr BaseCombinable(BaseCombinable const &other)
+      : parent{other.parent ? other.parent : &other},
+        identity{other.identity},
+        my_data{identity}
+  {
+  }
+
+  ~BaseCombinable()
+  {
+    if (parent && my_data != identity) {
```

Comment:
Justification for the additional check on identity? Is this to disallow extraneous folding with copies?

Reply:
It's because both destructors will get run when there's another outside, so it's to avoid extra folding yeah.